---
title: "riboprof_analysis_d14[20231005]"
output: html_notebook
---

Load all the required packages:
```{r}
library(tidyr)
library(dplyr)
library(stringr)
library(ggplot2)
library(riboWaltz)
library(data.table)
library(limma)
library(edgeR)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(GSA)

```

This notebook processes .bam files generated using STAR alignment against primary human GRCh38 GENCODE annotation.
.bam files used are in ./data/20230722_hiND14
processed files used are in ./data/hiND14v2

```{r}
# File contains gene counts for each transcript (rows) in each sample (columns)
combined_f <- './data/hiND14v2/combined_file.txt'
counts = read.table(combined_f, sep='\t', header=TRUE, row.names=1)
colnames(counts) =c('wt_rep1_RFP',
                    'wt_rep2_RFP',
                    'wt_rep3_RFP',
                    'Nterm_rep1_RFP',
                    'Nterm_rep2_RFP',
                    'Nterm_rep3_RFP',
                    'wt_rep1_RNA',
                    'wt_rep3_RNA',
                    'Nterm_rep2_RNA',
                    'wt_rep2_RNA',
                    'Nterm_rep3_RNA',
                    'Nterm_rep1_RNA')
head(counts)
```
```{r}
# convert to gene names
mart_path = './index/mart_homo-sapiens_geneID_transcriptID_transcriptID-v_gene-name.txt'
mart = read.table(mart_path, sep='\t', header=TRUE)
rownames(mart) = mart$Transcript.stable.ID

mart[1:10,]
```
```{r}
mart_gene.name.ID <- mart %>% mutate(Gene.name.ID = paste(Gene.name, Gene.stable.ID, sep = "_"))
gene.name.id_for_tid <- mart_gene.name.ID$Gene.name.ID
names(gene.name.id_for_tid) <- rownames(mart)
rownames(counts) = paste(gene.name.id_for_tid[rownames(counts)], rownames(counts), sep='_')

# name_for_tid = mart[,"Gene.name"]
# names(name_for_tid) = rownames(mart)
# rownames(counts) = paste(name_for_tid[rownames(counts)], rownames(counts), sep='_')

counts[1:10,]

dim(counts)
```

```{r}
# option to exclude mitochondria-encoded genes, also genes without gene names

counts = counts[-grep('^MT-*', rownames(counts)),]
counts = counts[grep('.+_', rownames(counts)),]

dim(counts)
counts[1:10,]
```
```{r}
# parse and assemble sample data

samples = array(data=NA, dim=c(ncol(counts), 3))
rownames(samples) = colnames(counts)
colnames(samples) = c('genotype', 'rep', 'RFP_or_RNA')

splits = strsplit(colnames(counts), '_')
for (i in 1:nrow(samples)) {
   samples[i,] = splits[[i]][1:3]
}

samples = data.frame(samples)

samples$RFP_or_RNA = factor(samples$RFP_or_RNA, levels=c('RFP', 'RNA'))
samples$genotype = factor(samples$genotype)
samples$rep = factor(substr(samples$rep, 4, 4), levels = c(1, 2, 3))
samples$grouping = paste(samples$genotype, samples$RFP_or_RNA, sep='_')

samples
```

### analysis for RNA with all 6 samples

```{r}
# first analyze only the RNA samples on their own

counts_rna = counts[,samples$RFP_or_RNA=='RNA']
samples_rna = samples[samples$RFP_or_RNA=='RNA',]
group_rna = samples_rna$grouping

data_rna = DGEList(counts = counts_rna, samples = samples_rna, group = group_rna)
```

```{r}
# filter out low-expressed genes
# try min.count = 10 (gene has to be expressed above 10 counts in a 'meaningful' number of samples)
# ^^ min.count = 10 is actually the default, I believe

# I think 'meaningful' is defined by the group attribute of data_rna

# the edgeR User's Guide mentions this should be performed on normalized counts,
# but in all the examples it's done before TMM normalization...

# going to generate a cpm version of data_rna and use that for determining which genes to keep.

data_rna_cpm <- cpm(data_rna, log=FALSE)

# generate the vector for which genes to keep

# if you don't want to use the filterByExpr function, can write your own
# criteria for which genes to keep

keep_cpm = filterByExpr(data_rna_cpm, min.count=10, group = data_rna$samples$group)
table(keep_cpm)

```

```{r}
# plot CPM histograms to compare raw vs. filtered data

L <- mean(data_rna$samples$lib.size) * 1e-6
M <- median(data_rna$samples$lib.size) * 1e-6

lcpm.cutoff <- log2(10/M + 2/L)
library(RColorBrewer)
nsamples <- ncol(data_rna)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))

lcpm <- cpm(data_rna, log=TRUE)

plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="A. Raw RNA data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
    den <- density(lcpm[,i])
    lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", rownames(data_rna$samples), text.col=col, bty="n")


# here is where I actually filter out the low-expressed genes from data_rna
data_rna = data_rna[keep_cpm, keep.lib.sizes=FALSE]

lcpm <- cpm(data_rna, log=TRUE)

plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="B. Filtered RNA data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
    den <- density(lcpm[,i])
    lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", rownames(data_rna$samples), text.col=col, bty="n")
```

```{r}
# calculate normalizing factors for RNA samples

data_rna = calcNormFactors(data_rna, method='TMM')

# plot multidimensional scaling -- an unsupervised look at the biggest
# differences between samples

lcpm = cpm(data_rna, log=TRUE)
plotMDS(lcpm, labels=rownames(data_rna$samples))
```

```{r}
plotMDS(lcpm, labels=rownames(data_rna$samples), dim.plot=c(3,4))
```
```{r}
data_rna$samples

line = data_rna$samples$genotype

# this design matrix is fairly simple. we intend to compare by 'line', and lane' is included as batch variable
design_rna = model.matrix(~0+line, data=data_rna$samples)
colnames(design_rna) = gsub('>', '_to_', colnames(design_rna))

design_rna
```
```{r}
# make contrasts

# rna_de = makeContrasts(linewt - lineNterm,
#                        levels=colnames(design_rna))
rna_de = makeContrasts(lineNterm - linewt,
                       levels=colnames(design_rna))

# perform voom

v_rna = voom(data_rna, design_rna, plot=TRUE)

# run the linear model

v_rna_fit = lmFit(v_rna, design_rna)
v_rna_fit = contrasts.fit(v_rna_fit, contrasts=rna_de)
e_rna_fit = eBayes(v_rna_fit)
plotSA(e_rna_fit, main='Final model: Mean-variance trend')

```
```{r}
# test for any significantly diff expr genes

dt_rna = decideTests(e_rna_fit, p.value=0.1)
summary(dt_rna)

# plots for mean expr vs logFC

plotMD(e_rna_fit, 
       column=1, 
       status=dt_rna[,1], 
       main=colnames(e_rna_fit)[1],
       ylim=c(-2,2))
```
```{r}
# generate a table of all kept genes as ranked by p-value
ranked_genes = topTable(e_rna_fit, coef=1, number=sum(keep_cpm), sort.by='p')

# volcano plot
ranked_genes$neglog10p = -log(ranked_genes$adj.P.Val, 10)
ranked_genes$plot_fill = rep('black', nrow(ranked_genes))
ranked_genes[ranked_genes$logFC > 0.25 & ranked_genes$adj.P.Val < 0.05,'plot_fill'] = 'purple'
ranked_genes[ranked_genes$logFC < -0.25 & ranked_genes$adj.P.Val < 0.05,'plot_fill'] = 'magenta'

pal = c('black', 'magenta', 'purple')

ranked_genes$plot_fill = factor(ranked_genes$plot_fill, levels = pal)

options(repr.plot.width=4, repr.plot.height=4)

ggplot(data=ranked_genes, aes(x=logFC, y=neglog10p, color=plot_fill)) + 
  geom_point() + 
  scale_color_manual(values=pal) + 
  theme_classic() + 
  theme(legend.position='none')

```
```{r}
# check distribution of p-values
hist(ranked_genes$P.Value)
```
### analysis for RFP w/ all 6 samples

```{r}
# only the RPF samples on their own

counts_rfp = counts[,samples$RFP_or_RNA=='RFP']
samples_rfp = samples[samples$RFP_or_RNA=='RFP',]
group_rfp = samples_rfp$grouping

data_rfp = DGEList(counts = counts_rfp, samples = samples_rfp, group = group_rfp)

# filter out low-expressed genes
# try min.count = 10 (gene has to be expressed above 10 counts in a 'meaningful' number of samples)
# ^^ min.count = 10 is actually the default, I believe

# I think 'meaningful' is defined by the group attribute of data_rna

# the edgeR User's Guide mentions this should be performed on normalized counts,
# but in all the examples it's done before TMM normalization...

# going to generate a cpm version of data_rna and use that for determining which genes to keep.

data_rfp_cpm <- cpm(data_rfp, log=FALSE)
head(data_rfp_cpm)

# generate the vector for which genes to keep

# if you don't want to use the filterByExpr function, can write your own
# criteria for which genes to keep

keep_cpm = filterByExpr(data_rfp_cpm, min.count=10, group = data_rfp$samples$group)
table(keep_cpm)
```
```{r}
# plot CPM histograms to compare raw vs. filtered data

L <- mean(data_rfp$samples$lib.size) * 1e-6
M <- median(data_rfp$samples$lib.size) * 1e-6

lcpm.cutoff <- log2(10/M + 2/L)
library(RColorBrewer)
nsamples <- ncol(data_rfp)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))

lcpm <- cpm(data_rfp, log=TRUE)

plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="A. Raw RFP data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
    den <- density(lcpm[,i])
    lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", rownames(data_rfp$samples), text.col=col, bty="n")


# here is where I actually filter out the low-expressed genes from data_rfp
data_rfp = data_rfp[keep_cpm, keep.lib.sizes=FALSE]

lcpm <- cpm(data_rfp, log=TRUE)

plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="B. Filtered RFP data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
    den <- density(lcpm[,i])
    lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", rownames(data_rfp$samples), text.col=col, bty="n")
```
```{r}
# calculate normalizing factors for RNA samples

data_rfp = calcNormFactors(data_rfp, method='TMM')

# plot multidimensional scaling -- an unsupervised look at the biggest
# differences between samples

lcpm = cpm(data_rfp, log=TRUE)
plotMDS(lcpm, labels=rownames(data_rfp$samples))
```
```{r}
plotMDS(lcpm, labels=rownames(data_rfp$samples), dim.plot=c(3,4))
```
```{r}
line = data_rfp$samples$genotype

# this design matrix is fairly simple. we intend to compare by 'line', and lane' is included as batch variable
design_rfp = model.matrix(~0+line, data=data_rfp$samples)
colnames(design_rfp) = gsub('>', '_to_', colnames(design_rfp))

design_rfp

# make contrasts

# rfp_de = makeContrasts(linewt - lineNterm,
#                        levels=colnames(design_rfp))
rfp_de = makeContrasts(lineNterm - linewt,
                       levels=colnames(design_rfp))

rfp_de
```
```{r}
# perform voom

v_rfp = voom(data_rfp, design_rfp, plot=TRUE)

# run the linear model

v_rfp_fit = lmFit(v_rfp, design_rfp)
v_rfp_fit = contrasts.fit(v_rfp_fit, contrasts=rfp_de)
e_rfp_fit = eBayes(v_rfp_fit)
plotSA(e_rfp_fit, main='Final model: Mean-variance trend')
```
```{r}
# test for any significantly diff expr genes

dt_rfp = decideTests(e_rfp_fit, adjust.method = 'BH', p.value=0.1)
summary(dt_rfp)

# plots for mean expr vs logFC

plotMD(e_rfp_fit, 
       column=1, 
       status=dt_rfp[,1], 
       main=colnames(e_rfp_fit)[1],
       ylim=c(-2,2))
```
```{r}
# generate a table of all kept genes as ranked by p-value
ranked_genes_rfp = topTable(e_rfp_fit, coef=1, number=sum(keep_cpm), sort.by='p')
```

### analyze differential TE for all 6 samples

```{r}
# create a DGE data object for all RNA and RPF samples

data_all = DGEList(counts = counts, samples = samples, group = samples$grouping)

# filter out low-expressed genes

# when calculating TE, Gerald filters by genes that are expressed at CPM > 1 in at least a min
# number of RPF libraries, b/c his RPF libraries have fewer reads...my RPF and RNA libraries are
# pretty similar in size...

# try min.count = 10 (gene has to be expressed above 10 counts in a 'meaningful' number of samples)
# ^^ min.count = 10 is actually the default, I believe

# I think 'meaningful' is defined by the group attribute of data_rna

# the edgeR User's Guide mentions this should be performed on normalized counts,
# but in all the examples it's done before TMM normalization...

# going to generate a cpm version of data_rna and use that for determining which genes to keep.

data_all_cpm <- cpm(data_all, log=FALSE)

# generate the vector for which genes to keep

keep_all_cpm = filterByExpr(data_all_cpm, min.count=10, data_all$samples$group)
table(keep_all_cpm)
```
```{r}
# plot CPM histograms to compare raw vs. filtered data

L <- mean(data_all$samples$lib.size) * 1e-6
M <- median(data_all$samples$lib.size) * 1e-6

lcpm.cutoff <- log2(10/M + 2/L)
library(RColorBrewer)
nsamples <- ncol(data_all)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))

lcpm <- cpm(data_all, log=TRUE)

plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="A. Raw RNA and RPF data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
    den <- density(lcpm[,i])
    lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", rownames(data_all$samples), text.col=col, bty="n")


# here is where I actually filter out the low-expressed genes from data_all
data_all = data_all[keep_all_cpm, keep.lib.sizes=FALSE]

lcpm <- cpm(data_all, log=TRUE)

plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="B. Filtered RNA and RPF data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
    den <- density(lcpm[,i])
    lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", rownames(data_all$samples), text.col=col, bty="n")
```
```{r}
# make subsets of data_all for just the RNA and ribo samples, so I can run TMM normalization on them separately.

data_rna_for_tmm = data_all[,samples$RFP_or_RNA=='RNA']
data_rna_for_tmm = calcNormFactors(data_rna_for_tmm, method='TMM')

data_rpf_for_tmm = data_all[,samples$RFP_or_RNA=='RFP']
data_rpf_for_tmm = calcNormFactors(data_rpf_for_tmm, method='TMM')

# recombine the separately calculated normalization factors into data_all
data_all$samples$norm.factors[samples$RFP_or_RNA=='RNA'] = data_rna_for_tmm$samples$norm.factors
data_all$samples$norm.factors[samples$RFP_or_RNA=='RFP'] = data_rpf_for_tmm$samples$norm.factors
```

```{r}
# Set TE
data_all_te <- data_all

# plot MDS -- not very informative since of course the 1st PC is RNA vs RPF

lcpm_all = cpm(data_all, log=TRUE)
plotMDS(lcpm, labels=rownames(data_all$samples))
```
```{r}
plotMDS(lcpm, labels=rownames(data_all$samples), dim.plot=c(3,4))
```
```{r}
# displaying again for easy reference in constructing design_all matrix
data_all$samples
```
```{r}
# our goal here is very similar to the example in the edgeR User's guide in section 3.5,
# 'Comparisons both between and within subjects'. The analogies are:
# Healthy/Disease1/Disease2 ~ S27>S27L, S27L>S27, and Unedited
# each patient ~ each replicate (note that in the example, they re-number their patients
# such that they have a patient #1, #2, and #3 in each disease)
# Treatment (hormone or none) (w/ a pair of treated/untreated samples for each patient) ~ RPF vs RNA 

# In the example they don't include the '0' so it's just ~Disease; I think it's more
# intuitive to include the 0 so it's explicitly clear what baseline is being subtracted
# however this only seems to solve this for the line variable, not lane or fraction...
# 

# line = factor(data_all$samples$genotype, levels=c('wt', 'Nterm'))
# fraction = factor(data_all$samples$RFP_or_RNA, levels=c('RNA', 'RFP'))
# 
# design_all = model.matrix(~0+line+fraction:line, data=data_all$samples)
# colnames(design_all) = gsub(':', '_', colnames(design_all))
# design_all

# make contrasts
# diff_te = makeContrasts(lineNterm_fractionRFP-linewt_fractionRFP,
#                         levels=colnames(design_all))
# 
# diff_te

line = factor(data_all$samples$genotype, levels=c('wt', 'Nterm'))
fraction = factor(data_all$samples$RFP_or_RNA, levels=c('RNA', 'RFP'))

design_all = model.matrix(~0+line:fraction, data=data_all$samples)
colnames(design_all) = gsub(':', '_', colnames(design_all))
design_all

# diff_te = makeContrasts(
#     wt_TE_vs_Nterm_TE = (linewt_fractionRFP-linewt_fractionRNA) - (lineNterm_fractionRFP-lineNterm_fractionRNA),
#                          levels=colnames(design_all))
diff_te = makeContrasts(
    Nterm_TE_vs_wt_TE = (lineNterm_fractionRFP-lineNterm_fractionRNA) - (linewt_fractionRFP-linewt_fractionRNA),
                         levels=colnames(design_all))


diff_te
```
```{r}
# perform voom

v_te = voom(data_all, design_all, plot=TRUE)

# note that the v_te$E matrix is pretty close to a log2(CPM + something)
# (online documentation calls this a 'normalized counts' matrix)
# I checked this by manually dividing the raw values in data_all$counts by data_all$samples$lib.size
# and taking log2 -- the results aren't identical to v_te$E but at least for the first 4x4 entries
# (which are all in the 2^5 to 2^8 range for CPM) they are within 0.01 (in log space)

v_te_fit = lmFit(v_te, design_all)
v_te_fit = contrasts.fit(v_te_fit, contrasts=diff_te)
e_te_fit = eBayes(v_te_fit)
plotSA(e_te_fit, main="Final model: Mean-variance trend")
```
```{r}
dt_te = decideTests(e_te_fit, p.value=0.1)
summary(dt_te)

head(e_te_fit)

#output_file <- "wtRFP-NtermRFP_plotMD_output.pdf"
#pdf(output_file)
plotMD(e_te_fit, column=1, status=dt_te[,1], main=colnames(e_te_fit)[1])
#dev.off()
```
```{r}
# volcano plot
top_te_genes = topTable(e_te_fit, coef=1, number=sum(keep_all_cpm), sort.by='p')

top_te_genes$neglog2p = -log(top_te_genes$adj.P.Val, 2)
top_te_genes$plot_fill = rep('black', nrow(top_te_genes))
top_te_genes[top_te_genes$logFC > 0.25 & top_te_genes$adj.P.Val < 0.05,'plot_fill'] = 'purple'
top_te_genes[top_te_genes$logFC < -0.25 & top_te_genes$adj.P.Val < 0.05,'plot_fill'] = 'magenta'

pal = c('black', 'magenta', 'purple')

top_te_genes$plot_fill = factor(top_te_genes$plot_fill, levels = pal)

options(repr.plot.width=4, repr.plot.height=4)

ggplot(data=top_te_genes, 
       aes(x=logFC, y=neglog2p, color=plot_fill)) + 
geom_point() + 
scale_color_manual(values=pal) + 
theme_classic() + 
theme(legend.position='none')
```
```{r}
head(top_te_genes)

# check distribution of p-values
hist(top_te_genes$P.Value)
```
```{r}
# check correlation between samples 
cor(v_te$E)
```

```{r}
# shortcut for finding the rowname corresponding to a gene
# EGR1
rownames(v_te$E)[grep('COL9A3', rownames(v_te$E))]
```

```{r}
library(repr)
options(repr.plot.width=4, repr.plot.height=4)

goi = 'COL9A3_ENSG00000092758_ENST00000649368'

# goi = 'LEF1_ENSG00000138795_ENST00000265165'

# goi = 'LLPH_ENST00000266604'
norm_counts = 2^v_te$E[goi,]

palette = c('purple1', 'purple4', 'magenta', 'magenta4', 'gray50', 'gray20')

df = data.frame(sample = names(norm_counts), cpm = norm_counts)
df = separate(df, sample, c("genotype", "replicate", "fraction"))
df$lines = factor(df$genotype, levels=c('wt', 'Nterm'))
df$xlabels = factor(paste(df$genotype, df$fraction, sep="_"),
                    levels=c('wt_RNA',
                             'wt_RFP', 
                             'Nterm_RNA', 
                             'Nterm_RFP'))

p1 = ggplot(df, aes(x=xlabels, y=cpm))
p1 + geom_point(aes(color = xlabels), size=2.5) + geom_path(group=df$xlabels) + ggtitle(goi) + theme_classic() + theme(legend.position = 'none', axis.text.x = element_text(angle = 90)) + expand_limits(y=0) + scale_color_manual(values = palette)
```

Redo the graph:

```{r}
library(repr)
options(repr.plot.width=4, repr.plot.height=4)

# shortcut for finding the rowname corresponding to a gene
# EGR1
rownames(v_te$E)[grep('CDH1', rownames(v_te$E))]

# goi = c('COL9A3_ENSG00000092758_ENST00000649368',
#         'CDH1_ENSG00000039068_ENST00000261769',
#         'LEF1_ENSG00000138795_ENST00000265165')
# goi = 'LLPH_ENST00000266604'

# norm_counts = 2^v_te$E[goi,]

df_wide <- as_tibble(2^v_te$E) %>% mutate(gene = rownames(2^v_te$E), .before=1)
df_tall <- gather(df_wide, key = "sample", value = "cpm", -gene)
df_tall <- separate(df_tall, sample, c("genotype", "replicate", "fraction"))

df_tall_te <- spread(df_tall, key = fraction, value = cpm)
df_tall_te$TE <- df_tall_te$RFP/df_tall_te$RNA
df_tall_te <- separate(df_tall_te, gene, c("gene_name", "ensembl_gene", "transcript_id"),sep = "_", remove = FALSE, extra = "drop")
df_tall_te_only <- df_tall_te %>% 
  dplyr::select(-RNA, -RFP) %>% 
  pivot_wider(names_from = replicate, values_from = TE) %>% 
  rowwise() %>% 
  mutate(ave_te = mean(c(rep1, rep2, rep3), na.rm = TRUE),
         std_te = sd(c(rep1, rep2, rep3), na.rm = TRUE))
summary <- df_tall_te_only %>% 
  pivot_longer(cols = c(rep1, rep2, rep3), names_to = c("replicate"), values_to = "values")

# Create an empty data frame to store the t-test results
t_test_results <- data.frame(transcript_id = character(0),
                              p_value = numeric(0))

# Perform t-test for each 'transcript_id'
for (tid in unique(summary$transcript_id)) {
  subset_data <- summary %>% filter(tid == .data$transcript_id)
  t_test_result <- t.test(values ~ genotype, data = subset_data)
  t_test_results <- rbind(t_test_results, 
                          data.frame(transcript_id = tid, p_value = t_test_result$p.value))
}

summary_ttest <- summary %>% 
  dplyr::select(-values, -replicate) %>% 
  distinct() %>% 
  pivot_wider(names_from = genotype, values_from = c(ave_te, std_te)) %>% 
  mutate(diff_te = ave_te_Nterm/ave_te_wt,
         downregulated = ifelse(diff_te > 1, 0, 1)) %>% 
  full_join(t_test_results, by = "transcript_id")
summary_ttest <- summary_ttest %>% 
  arrange(p_value)

# write.csv(summary_ttest, file = "TE_with_ttest.csv", row.names = FALSE)

# df_tall$lines = factor(df_tall$genotype, levels=c('wt', 'Nterm'))
# df_tall$xlabels = factor(paste(df_tall$genotype, df_tall$fraction, sep="_"),
#                          levels=c('wt_RNA', 'wt_RFP', 'Nterm_RNA', 'Nterm_RFP'))

#palette = c('purple1', 'purple4', 'magenta', 'magenta4', 'gray50', 'gray20')

# subset_tids <- c('CSPG4_ENSG00000173546_ENST00000308508',
#                  'RHOBTB2_ENSG00000008853_ENST00000251822',
#                  'FNTA_ENSG00000168522_ENST00000302279',
#                  'MYORG_ENSG00000164976_ENST00000297625',
#                  'PABPC5_ENSG00000174740_ENST00000312600')

# subset_tids <- c('PCDHB5_ENSG00000113209_ENST00000231134',
#                  'CDH5_ENSG00000179776_ENST00000341529',
#                  'PCOLCE2_ENSG00000163710_ENST00000295992',
#                  'PCDHA10_ENSG00000250120_ENST00000307360',
#                  'RAPGEF6_ENSG00000158987_ENST00000509018')

subset_tids <- c('CSPG4_ENSG00000173546_ENST00000308508',
                 'RHOBTB2_ENSG00000008853_ENST00000251822',
                 'TTC23_ENSG00000103852_ENST00000394132',
                 'MYORG_ENSG00000164976_ENST00000297625',
                 'SEMA3A_ENSG00000075213_ENST00000265362')

plot <- summary %>% 
  dplyr::filter(gene %in% subset_tids)
plot$genotype <- factor(plot$genotype, levels=c('wt', 'Nterm'))

custom_fill <- c("gray80", "gray40")

ggplot(plot, aes(x=genotype, y=ave_te, fill = genotype)) + 
  geom_bar(stat = "identity", position = "dodge") +
  geom_errorbar(
    aes(ymin = ave_te - std_te, ymax = ave_te + std_te),
    position = position_dodge(width = 0.8),
    width = 0.2
  ) +
  facet_wrap (~ gene_name, scales = "free_y", ncol = 5) +
  scale_fill_manual(values = custom_fill) +
  theme_classic() + 
  theme(legend.position = 'none', axis.text.x = element_text(angle = 90)) +
  expand_limits(y=0)

# ggsave("ecm_genes_downregulated_nterm.pdf", width=12, height=9)
ggsave("disease_genes_downregulated_nterm.pdf", width=12, height=9)

# 

```




```{r}
# Scatter plot of rna and rfp sig. genes
# keep_rfp_cpm = filterByExpr(data_rfp_cpm, min.count=10, group = data_rfp$samples$group)
# keep_rna_cpm = filterByExpr(data_rna_cpm, min.count=10, group = data_rna$samples$group)
# 
# toptable_rfp = topTable(e_rfp_fit, coef=1, number=sum(keep_rfp_cpm), sort.by='p')
# toptable_rna = topTable(e_rfp_fit, coef=1, number=sum(keep_rna_cpm), sort.by='p')

# rfp_fc_sig <- tibble("gene" =  row.names(toptable_rfp),
#                      "RFP_Nterm-wt_fc" = toptable_rfp$logFC,
#                      "RFP_Nterm-wt_sig" = toptable_rfp$adj.P.Val)
# rna_fc_sig <- tibble("gene" = row.names(toptable_rna),
#                      "RNA_Nterm-wt_fc" = toptable_rna$logFC,
#                      "RNA_Nterm-wt_sig" = toptable_rna$adj.P.Val)
# all_fc_sig <- full_join(rfp_fc_sig, rna_fc_sig, by = "gene") %>% 
#   separate(gene, into = c("gene_name", "geneID", "transcriptID"), sep = "_", remove = FALSE)
# all_fc_sig$entrez <- mapIds(org.Hs.eg.db, keys=all_fc_sig$geneID, column="ENTREZID", keytype="ENSEMBL", multiVals="first")
# 
# # 0 is not sig, 1 is only sig in rna, 2 is only sig in rfp, 3 is sig in both.
# all_fc_sig$group <- factor(1 * (all_fc_sig$`RNA_Nterm-wt_sig` < 0.1) + 2 * (all_fc_sig$`RFP_Nterm-wt_sig` < 0.1))



rfp_fc_sig <- tibble("gene" =  row.names(e_rfp_fit$coefficients),
                     "RFP_Nterm-wt_fc" = e_rfp_fit$coefficients[,1],
                     "RFP_Nterm-wt_sig" = as.vector(dt_rfp),
                     "RFP_Nterm-wt_FDR" = p.adjust(e_rfp_fit$p.value, method = "BH", n = length(e_rfp_fit$p.value)))
rna_fc_sig <- tibble("gene" = row.names(e_rna_fit$coefficients),
                     "RNA_Nterm-wt_fc" = e_rna_fit$coefficients[,1],
                     "RNA_Nterm-wt_sig" = as.vector(dt_rna),
                     "RNA_Nterm-wt_FDR" = p.adjust(e_rna_fit$p.value, method = "BH", n = length(e_rna_fit$p.value)))
all_fc_sig <- full_join(rfp_fc_sig, rna_fc_sig, by = "gene") %>%
  separate(gene, into = c("gene_name", "geneID", "transcriptID"), sep = "_", remove = FALSE)
all_fc_sig$entrez <- mapIds(org.Hs.eg.db, keys=all_fc_sig$geneID, column="ENTREZID", keytype="ENSEMBL", multiVals="first")

# 0 is not sig, 1 is only sig in rna, 2 is only sig in rfp, 3 is sig in both.
all_fc_sig$group <- factor(
  1 * (!is.na (all_fc_sig$`RNA_Nterm-wt_sig`) & all_fc_sig$`RNA_Nterm-wt_sig` != 0) +
    2 * (!is.na (all_fc_sig$`RFP_Nterm-wt_sig`) & all_fc_sig$`RFP_Nterm-wt_sig` != 0)
)

# dt_rfp = decideTests(e_rfp_fit, adjust.method = 'BH', p.value=0.1)

palette = c('black', 'blue', 'red', 'magenta4')
legends = c('no change', 'transcriptional changes', 'translational changes', 'both')

ggplot(all_fc_sig, aes(x=`RNA_Nterm-wt_fc`, y=`RFP_Nterm-wt_fc`)) +
  geom_point(aes(color = group), size = 1) +
  scale_color_manual(
    values = palette,
    labels = legends) +
  theme_minimal() +
  xlim(-2, 2) + ylim (-2, 2) +
  xlab("Log2 FC transcriptome") + ylab("Log2 FC translatome")

write.csv(all_fc_sig, file = "all_fc_sig_hiNs_d14.csv", quote = FALSE, sep = ",", row.names = FALSE)
# ggsave("rna_rfp_sig_genes.pdf", width=12, height=9)

```


```{r}
translation_up <- all_fc_sig %>% filter(group == 2 & `RFP_Nterm-wt_sig` == 1) %>% dplyr::select(gene_name)
translation_down <- all_fc_sig %>% filter(group == 2 & `RFP_Nterm-wt_sig` == -1) %>% dplyr::select(gene_name)
all_genes <- all_fc_sig %>% dplyr::select(gene_name)

# write.csv(translation_up, file = "./output/hiND14_translation_up.csv", row.names = FALSE)
# write.csv(translation_down, file = "./output/hiND14_translation_down.csv", row.names = FALSE)
# write.csv(all_genes, file = "./output/hiND14_all_genes.csv", row.names = FALSE)

```

```{r}
rna_only <- all_fc_sig %>% filter(group == 1) %>% dplyr::select(gene)

# backup_v_rfp_weights <- v_rfp$weights
# backup_v_rfp_E <- v_rfp$E

matching_indices_rna_only <- which(rownames(v_rfp$E) %in% rna_only$gene)
temp_v_rfp_E <- v_rfp$E[-matching_indices_rna_only, ]
temp_v_rfp_weights <- v_rfp$weights[-matching_indices_rna_only, ]

v_rfp$E <- temp_v_rfp_E
v_rfp$weights <- temp_v_rfp_weights
```


```{r}
# all_fc_sig$entrez <- mapIds(org.Hs.eg.db, keys=all_fc_sig$geneID, column="ENTREZID", keytype="ENSEMBL", multiVals="first")
# 
# gmt_file <- "./data/Human_GOBP_AllPathways_no_GO_iea_October_01_2023_entrezgene.gmt"
# genesets <- GSA.read.gmt(gmt_file)
# names(genesets$genesets) <- genesets$geneset.names
# genesets_filtered <- ids2indices(genesets$genesets,id=all_fc_sig$entrez, remove.empty=TRUE)
# geneset_sizes <- unlist(lapply(genesets_filtered, length))
# geneset_indices <- which(geneset_sizes>=10 & geneset_sizes<500)
# 
# # v_rfp$E <- backup_v_rfp_E
# camera_results <- camera(v_rfp, genesets_filtered[geneset_indices], design_rfp, contrast=rfp_de[,1])
# # camera_results <- camera(v_te, genesets_filtered[geneset_indices], design_all, contrast=diff_te[,1])
# camera_descr <- unlist(lapply(rownames(camera_results), function(x){unlist(strsplit(x,"\\%"))[1]}))
# camera_Phenotype <- unlist(lapply(camera_results[,"Direction"], function(x){if(x=="Up"){1}else{(-1)}}))
# camera_genes <- c()
# 
# # camera_results_file <- paste("camera", "_", "v_rfp_d14",
# #                              "_", "Human_GOBP_AllPathways_no_GO_iea_October_01_2023_entrezgene", ".txt", sep="")
# camera_results_file <- paste("camera", "_", "v_rfp_d14_nona_removeRNA",
#                              "_", "Human_GOBP_AllPathways_no_GO_iea_October_01_2023_entrezgene", ".txt", sep="")
# 
# # camera_results_file <- paste("camera", "_", "v_te_d14", 
# #                              "_", "Human_GOBP_AllPathways_no_GO_iea_October_01_2023_entrezgene", ".txt", sep="")
# 
# 
# for(j in 1:length(rownames(camera_results))){
#   current_geneset <- unlist(genesets_filtered[which(names(genesets_filtered) %in% 
#                                                       rownames(camera_results)[j])])
#   current_genes <- c()
#   for(k in 1:length(current_geneset)){
#     if(k==length(current_geneset)){
#       current_genes <- paste(current_genes, all_fc_sig$entrez[current_geneset[k]], sep="")
#     }else{
#       current_genes <- paste(current_genes, all_fc_sig$entrez[current_geneset[k]], ",", sep="")
#     }
#   }
#   camera_genes <- rbind(camera_genes, current_genes)
# }
# 
# rownames(camera_genes) <- rownames(camera_results)
# camera_results_generic_em <- data.frame(rownames(camera_results),
#                                         camera_descr,
#                                         PValue = camera_results[,"PValue"],
#                                         FDR=camera_results[,"FDR"],
#                                         camera_Phenotype,
#                                         camera_genes)
# # write.table(camera_results_generic_em, camera_results_file, col.name=TRUE, sep="\t", row.names=FALSE, quote=FALSE)
# 
# camera_results_generic_em_nona <- na.omit(camera_results_generic_em)
# write.table(camera_results_generic_em_nona, camera_results_file, col.name=TRUE, sep="\t", row.names=FALSE, quote=FALSE)
# 

```

```{r}

# #use easy cyRest library to communicate with cytoscape.
# 
# tryCatch(expr = { library(RCy3)}, 
#          error = function(e) { install_github("cytoscape/RCy3")}, finally = library(RCy3))
# 
# #defined threshold for GSEA enrichments (need to be strings for cyrest call)
# pvalue_threshold <- "0.05"
# qvalue_threshold <- "0.001"
# 
# similarity_threshold <- "0.25"
# similarity_metric = "JACCARD"
# 
# generic_gmt_file <- file.path(getwd(),gmt_file)
# 
# analysis_name <- "Nterm_vs_wt"
# cur_model_name <- paste("camera",analysis_name,sep="_")
# results_filename <- file.path(getwd(),camera_results_file)
# 
# #######################################
# #create EM - camera results
# #######################################
# current_network_name <- paste(cur_model_name,pvalue_threshold,qvalue_threshold,sep="_")
# 
# em_command = paste('enrichmentmap build analysisType="generic"',
#                    'gmtFile=',generic_gmt_file,
#                    'pvalue=',pvalue_threshold,
#                    'qvalue=',qvalue_threshold,
#                    'similaritycutoff=',similarity_threshold,
#                    'coefficients=',similarity_metric,
#                    'enrichmentsDataset1=',results_filename,
#                    # 'expressionDataset1=',file.path(getwd(),
#                    #                             expression_file),
#                    sep=" ")
# 
# #enrichment map command will return the suid of newly created network.
# response <- commandsGET(em_command)
# 
# 
# current_network_suid <- 0
# #enrichment map command will return the suid of newly created network unless it Failed.  
# #If it failed it will contain the word failed
# if(grepl(pattern="Failed", response)){
#   paste(response)
# } else {
#   current_network_suid <- response
# }
# 
# response <- renameNetwork(current_network_name, as.numeric(current_network_suid))

```



### riboWaltz analysis

I am using riboWaltz to process this data.
I create annotation_dt using the human primary assembly basic annotation that I used to align my RFP data with STAR. I'm also loading the data and convert them into list.

```{r}
annotation_dt=create_annotation(gtfpath = "ref/gencode.v43.primary_assembly.basic.annotation.gtf")
bam_list <- bamtolist("data/20230722_hiND14", annotation = annotation_dt, refseq_sep = "|")
```
Call for psite, flanking = 6, extremity = "auto"
Use default recommended settings:
Also save the offsets per sample in a text file

```{r}
offsets <- psite(data = bam_list)
samples <- unique(offsets$sample)
for (i in 1:length(samples)) {
  sam_offs <- subset(offsets, sample == samples[i], 
                    select = c("length", "corrected_offset_from_3"))
  colnames(sam_offs) <- c("length", "p_offset")
  write.table(sam_offs, paste("output/20230722_hiND14/", samples[i], "_p-site-offsets.tsv", sep=""), 
              sep = "\t", row.names = FALSE)
}
```

Get the main table of reads with psite information.
Tried saving each psite list for each sample, but the filesize is way too big. Don't do it.

```{r}
reads_psite_list <- psite_info(bam_list, offsets,
                               fastapath = "./ref/GRCh38.primary_assembly.genome.fa", 
                               fasta_genome = TRUE, refseq_sep = " ",
                               gtfpath = "./ref/gencode.v43.primary_assembly.basic.annotation.gtf")
```

Given that I'm interested to look at a-site codon usage, I'm filtering out all reads mapped to 3'utr, and those mapped right before the stop codon. 
A lot of those reads threw an error unfortunately. The package is being updated

```{r}
for(n in names(reads_psite_list)) {
  reads_psite_list[[n]] <- reads_psite_list[[n]][!which(reads_psite_list[[n]]$psite_region == "3utr")]
  reads_psite_list[[n]] <- reads_psite_list[[n]][!which(reads_psite_list[[n]]$psite_region == "cds" & reads_psite_list[[n]]$end3 < reads_psite_list[[n]]$psite + 3)]
  reads_psite_list[[n]] <- reads_psite_list[[n]][!is.na(reads_psite_list[[n]]$psite_region)]
  reads_psite_list[[n]][, c("stable_id", "version") := tstrsplit(transcript, "\\.")]
}
```

Subsetting transcripts that are translationally changed:
p-site
```{r}
translation_tID <- all_fc_sig %>% filter(group == 2) %>% dplyr::select(transcriptID)
translation_up_tID <- all_fc_sig %>% filter(group == 2 & `RFP_Nterm-wt_sig` == 1) %>% dplyr::select(transcriptID)
translation_down_tID <- all_fc_sig %>% filter(group == 2 & `RFP_Nterm-wt_sig` == -1) %>% dplyr::select(transcriptID)
control_tID <- all_fc_sig %>% filter(group == 1 | group == 0) %>% dplyr::select(transcriptID)
black_control_tID <- all_fc_sig %>% filter(group == 0) %>% dplyr::select(transcriptID)

control_df <- reads_psite_list$hiND14_Nterm_rep_1_RFP[
  reads_psite_list$hiND14_Nterm_rep_1_RFP$stable_id %in% control_tID$transcriptID, ]
# Keep only unique stable_id
unique_control_df <- unique(control_df[, .(cds_length = cds_stop-cds_start), by = stable_id]) %>% 
  mutate(sample = "translation_unchanged")

black_control_df <- reads_psite_list$hiND14_Nterm_rep_1_RFP[
  reads_psite_list$hiND14_Nterm_rep_1_RFP$stable_id %in% black_control_tID$transcriptID, ]
# Keep only unique stable_id
unique_black_control_df <- unique(control_df[, .(cds_length = cds_stop-cds_start), by = stable_id]) %>% 
  mutate(sample = "translation_unchanged_black_only")



# Check for those upregulated in translation
# I have to use data.frame to handle large datasets like this.
translation_up_df <- reads_psite_list$hiND14_Nterm_rep_1_RFP[
  reads_psite_list$hiND14_Nterm_rep_1_RFP$stable_id %in% translation_up_tID$transcriptID, ]
# Keep only unique stable_id
unique_translation_up_df <- unique(translation_up_df[, .(cds_length = cds_stop-cds_start), by = stable_id]) %>% 
  mutate(sample = "translation_up")

# Check for those downregulated in translation
translation_down_df <- reads_psite_list$hiND14_Nterm_rep_1_RFP[
  reads_psite_list$hiND14_Nterm_rep_1_RFP$stable_id %in% translation_down_tID$transcriptID, ]
unique_translation_down_df <- unique(translation_down_df[, .(cds_length = cds_stop-cds_start), by = stable_id]) %>% 
  mutate(sample = "translation_down")

plot_df <- rbind(unique_control_df, unique_translation_up_df, unique_translation_down_df)
plot_df$sample <- factor(plot_df$sample, levels = c("translation_unchanged", "translation_up", "translation_down"))

ggplot(plot_df, aes(x = sample, y = cds_length)) +
  geom_boxplot(width = 0.4, position = position_dodge(width = 0.5)) +
  labs(x = "sample", y = "cds_length") +
  theme_classic()

  # ylim(0,6000) +

ggsave("translation_cds_length_6000.pdf", width = 6, height = 4)

wilcox.test(unique_control_df$cds_length, unique_translation_up_df$cds_length)
wilcox.test(unique_translation_down_df$cds_length, unique_control_df$cds_length)
wilcox.test(unique_translation_down_df$cds_length, unique_translation_up_df$cds_length)





# reads_psite_split_list <- list()
# 
# dt <- separate(reads_psite_list$hiND14_Nterm_rep_1_RFP, transcript, into = c("transcript", "version"), sep = ".")
# for(name in names(reads_psite_list)) {
#   tibble_data <- separate(reads_psite_list[[name]], transcript,
#                           into = c("transcript", "version"), sep = ".")
#   reads_psite_split_list[[name]] <- tibble_data
# }  
```


Checking for codon usage:
p-site

```{r}
fasta_path = "ref/GRCh38.primary_assembly.genome.fa"
gtf_path = "ref/gencode.v43.primary_assembly.basic.annotation.gtf"
codon_table <- read.table(file = "codon_table.tsv", sep = "\t", header=TRUE)

codon_usage_psite_all <- codon_table
codon_rank_psite_all <- codon_table

for(n in names(reads_psite_list)) {
  codon_usage <- codon_usage_psite(reads_psite_list, annotation = annotation_dt, 
                                   sample = n, fastapath = fasta_path, gtfpath = gtf_path, 
                                   fasta_genome = TRUE, refseq_sep = " ", site = "psite")
  codon_usage_psite_all <- left_join(codon_usage_psite_all,
                                     codon_usage$dt[, .(codon, normalized_value)], by="codon")
  normalized_value_index <- which(colnames(codon_usage_psite_all) == "normalized_value")
  colnames(codon_usage_psite_all)[normalized_value_index] <- n
  codon_rank_psite_all[[n]] <- rank(codon_usage_psite_all[[n]])
}

codon_usage_psite_all$wt_RFP_ave = rowMeans(
  codon_usage_psite_all[, c("hiND14_wt_rep_1_RFP", "hiND14_wt_rep_2_RFP", "hiND14_wt_rep_3_RFP")])
codon_usage_psite_all$Nterm_RFP_ave = rowMeans(
  codon_usage_psite_all[, c("hiND14_Nterm_rep_1_RFP", "hiND14_Nterm_rep_2_RFP", "hiND14_Nterm_rep_3_RFP")])
codon_usage_psite_all <- codon_usage_psite_all %>%
  mutate(wt_RFP_ave_plot = wt_RFP_ave/max(codon_usage_psite_all$wt_RFP_ave),
         Nterm_RFP_ave_plot = Nterm_RFP_ave/max(codon_usage_psite_all$Nterm_RFP_ave))
```

Trying to do wilcoxon rank-sum test:

```{r}
wt <- rowMeans(codon_rank_psite_all[, c("hiND14_wt_rep_1_RFP", "hiND14_wt_rep_2_RFP", "hiND14_wt_rep_3_RFP")])
Nterm <- rowMeans(codon_rank_psite_all[, c("hiND14_Nterm_rep_1_RFP", "hiND14_Nterm_rep_2_RFP", "hiND14_Nterm_rep_3_RFP")])
wilcox.test(wt, Nterm)

# No change in terms of codon rank usage.
```

a-site.

```{r}
fasta_path = "data/Try2/GRCh38.primary_assembly.genome.fa"
gtf_path = "data/Try2/gencode.v43.primary_assembly.basic.annotation.gtf"
codon_table <- read.table(file = "codon_table.tsv", sep = "\t", header=TRUE)

codon_usage_asite_all <- codon_table
codon_rank_asite_all <- codon_table

for(n in names(reads_psite_list)) {
  codon_usage <- codon_usage_psite(reads_psite_list, annotation = annotation_dt, 
                                   sample = n, fastapath = fasta_path, gtfpath = gtf_path, 
                                   fasta_genome = TRUE, refseq_sep = " ", site = "asite")
  codon_usage_asite_all <- left_join(codon_usage_asite_all,
                                     codon_usage$dt[, .(codon, normalized_value)], by="codon")
  normalized_value_index <- which(colnames(codon_usage_asite_all) == "normalized_value")
  colnames(codon_usage_asite_all)[normalized_value_index] <- n
  codon_rank_asite_all[[n]] <- rank(codon_usage_asite_all[[n]])
}

codon_usage_asite_all$wt_RFP_ave = rowMeans(
  codon_usage_asite_all[, c("hiND14_wt_rep_1_RFP", "hiND14_wt_rep_2_RFP", "hiND14_wt_rep_3_RFP")])
codon_usage_asite_all$Nterm_RFP_ave = rowMeans(
  codon_usage_asite_all[, c("hiND14_Nterm_rep_1_RFP", "hiND14_Nterm_rep_2_RFP", "hiND14_Nterm_rep_3_RFP")])
codon_usage_asite_all <- codon_usage_asite_all %>%
  mutate(wt_RFP_ave_plot = wt_RFP_ave/max(codon_usage_asite_all$wt_RFP_ave),
         Nterm_RFP_ave_plot = Nterm_RFP_ave/max(codon_usage_asite_all$Nterm_RFP_ave))
```
Trying to do wilcoxon rank-sum test:

```{r}
wt <- rowMeans(codon_rank_asite_all[, c("hiND14_wt_rep_1_RFP", "hiND14_wt_rep_2_RFP", "hiND14_wt_rep_3_RFP")])
Nterm <- rowMeans(codon_rank_asite_all[, c("hiND14_Nterm_rep_1_RFP", "hiND14_Nterm_rep_2_RFP", "hiND14_Nterm_rep_3_RFP")])
wilcox.test(wt, Nterm)

# No change in terms of codon rank usage.
```

Do a scatter plot with linear model fitting

```{r}
bs <- 20
label_number <- 0
# plot_df <- codon_usage_psite_all
plot_df <- codon_usage_asite_all
setDT(plot_df)
plot_df <- plot_df[order(wt_RFP_ave_plot)][, `:=`(codon, factor(codon, levels = codon))]

correlation <- round(cor(plot_df$wt_RFP_ave_plot, 
                         plot_df$Nterm_RFP_ave_plot), 3)

fit <- lm(plot_df$Nterm_RFP_ave_plot ~ plot_df$wt_RFP_ave_plot)
outlier_tab <- data.table(predict.lm(fit, interval = "confidence", level = 0.95))[, `:=`(codon, plot_df$codon)][, `:=`(Nterm_RFP_ave_plot, plot_df$Nterm_RFP_ave_plot)][Nterm_RFP_ave_plot < lwr | Nterm_RFP_ave_plot > upr]
outlier_cod <- as.character(outlier_tab[, `:=`(max_dist, min(abs(Nterm_RFP_ave_plot - lwr), 
                                                             abs(Nterm_RFP_ave_plot - upr))), 
                                        by = 1:nrow(outlier_tab)][order(-max_dist)][1:label_number, codon])
lab_table <- plot_df[, `:=`(codon_out, "")][as.character(codon) %in% outlier_cod, `:=`(codon_out, codon)
                                               ][, `:=`(aa_out, "")][codon_out != "", `:=`(aa_out, aa)]

ggplot(plot_df, aes(x = wt_RFP_ave_plot, y=Nterm_RFP_ave_plot)) +
  geom_smooth(method = "lm", se = TRUE, color = "gray80", fill = "gray80", linetype= 1,
              formula = y ~ x, level = 0.95, fullrange = TRUE) +
  geom_point(alpha = 0.5, size = bs* 0.14) +
  theme_bw(base_size = bs) +
  theme(legend.position = "top", legend.margin = margin(0,0,0,0),
        legend.box.margin = margin(5,0,-15,0)) +
  theme(legend.text = element_text(margin = margin(l = -12, unit = "pt"))) +
  theme(panel.grid.minor.y = element_blank(), panel.grid.minor.x = element_blank()) +
  scale_x_continuous(name = "LLPH wt", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  scale_y_continuous(name = "LLPH Nterm", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  coord_cartesian(xlim = c(-0.05, 1.05), ylim = c(-0.05, 1.05)) +
  annotate("text", x = 1, y = 0, label = paste0("R=", correlation), 
           vjust = -0.2, size = bs * 0.2, hjust = 1, color = "black") +
  geom_abline(intercept = 0, slope = 1, color = "red") +
  ggrepel::geom_text_repel(data = lab_table, aes(label = codon_out), show.legend = F)

#ggsave("Nterm_vs_wt_codon_psite.pdf")
ggsave("Nterm_vs_wt_codon_asite.pdf")

```
```{r}
# Sample data
x <- plot_df$wt_RFP_ave_plot
y <- plot_df$Nterm_RFP_ave_plot

# Fit a linear regression model
model <- lm(y ~ x)

# Test for a different slope
slope_test <- summary(model)$coefficients[2, 4]  # p-value for the slope coefficient

# Test for a different intercept
intercept_test <- summary(model)$coefficients[1, 4]  # p-value for the intercept coefficient

# Set the significance level
alpha <- 0.05

# Test the hypotheses
if (slope_test < alpha) {
  cat("Reject H0 for slope: The slope of the regression line is different from y = x.\n")
} else {
  cat("Fail to reject H0 for slope: There is no significant difference in slope from y = x.\n")
}

if (intercept_test < alpha) {
  cat("Reject H0 for intercept: The intercept of the regression line is different from y = x.\n")
} else {
  cat("Fail to reject H0 for intercept: There is no significant difference in intercept from y = x.\n")
}

# Print the p-values
cat("p-value for slope:", slope_test, "\n")
cat("p-value for intercept:", intercept_test, "\n")

```





Try between wt, check for a-site:

```{r}
bs <- 25
label_number <- 10
# plot_df <- codon_usage_psite_all
plot_df <- codon_usage_asite_all
setDT(plot_df)
plot_df <- plot_df[order(hiND14_Nterm_rep_1_RFP)][, `:=`(codon, factor(codon, levels = codon))]

correlation <- round(cor(plot_df$hiND14_Nterm_rep_1_RFP, 
                         plot_df$hiND14_Nterm_rep_2_RFP), 3)

fit <- lm(plot_df$hiND14_Nterm_rep_2_RFP ~ plot_df$hiND14_Nterm_rep_1_RFP)
outlier_tab <- data.table(predict.lm(fit, interval = "confidence", level = 0.95))[, `:=`(codon, plot_df$codon)][, `:=`(hiND14_Nterm_rep_2_RFP, plot_df$hiND14_Nterm_rep_2_RFP)][hiND14_Nterm_rep_2_RFP < lwr | hiND14_Nterm_rep_2_RFP > upr]
outlier_cod <- as.character(outlier_tab[, `:=`(max_dist, min(abs(hiND14_Nterm_rep_2_RFP - lwr), 
                                                             abs(hiND14_Nterm_rep_2_RFP - upr))), 
                                        by = 1:nrow(outlier_tab)][order(-max_dist)][1:label_number, codon])
lab_table <- plot_df[, `:=`(codon_out, "")][as.character(codon) %in% outlier_cod, `:=`(codon_out, codon)
                                               ][, `:=`(aa_out, "")][codon_out != "", `:=`(aa_out, aa)]

ggplot(plot_df, aes(x = hiND14_Nterm_rep_1_RFP, y=hiND14_Nterm_rep_2_RFP)) +
  geom_smooth(method = "lm", se = TRUE, color = "gray80", fill = "gray80", linetype= 1,
              formula = y ~ x, level = 0.95, fullrange = TRUE) +
  geom_point(alpha = 0.5, size = bs* 0.14) +
  theme_bw(base_size = bs) +
  theme(legend.position = "top", legend.margin = margin(0,0,0,0),
        legend.box.margin = margin(5,0,-15,0)) +
  theme(legend.text = element_text(margin = margin(l = -12, unit = "pt"))) +
  theme(panel.grid.minor.y = element_blank(), panel.grid.minor.x = element_blank()) +
  annotate("text", x = 1, y = 0, label = paste0("R=", correlation), 
           vjust = -0.2, size = bs * 0.2, hjust = 1, color = "black") +
  geom_abline(intercept = 0, slope = 1, color = "red") +
  ggrepel::geom_text_repel(data = lab_table, aes(label = codon_out), show.legend = F)

```


Check how the reads are distributed for a given transcript.

```{r}
# To plot read counts against position given a particular gene in a given sample:
transcript_id = "ENST00000239938.5"
reads_psite_sample = "hiND14_Nterm_rep_3_RFP"

# Get the read counts for a given p-siste position of a specific transcript in a given sample.
psite_counts <- reads_psite_list[[reads_psite_sample]][
  reads_psite_list[[reads_psite_sample]]$transcript == transcript_id ,
  ]$psite

# Convert counts table to data frame, change the column names.
counts <- as.data.frame(table(psite_counts))
colnames(counts) <- c("Position", "ReadCount")

# Generate a sequence of numbers from 0 or min. number to the max. number.
counts$Position <- as.numeric(as.character(counts$Position))
number_range <- seq(pmin(min(counts$Position), 0), max(counts$Position))

# Complete the data frame with missing numbers and set their counts to 0.
# Normalize ReadCount with the size of the library.
nrow_reads = nrow(reads_psite_list[[reads_psite_sample]])
counts_complete <- counts %>%
  complete(Position = number_range) %>%
  replace_na(list(ReadCount = 0)) %>%
  mutate(ReadCount = ReadCount / nrow_reads * 1000000)

# Parse information about genotype, rfp/rna, and rep number.
genotype <- ""
if (grepl("wt", reads_psite_sample)) genotype = "wt" else genotype = "Nterm"
rfp_or_rna  <- ""
if (grepl("RFP", reads_psite_sample)) rfp_or_rna = "RFP" else rfp_or_rna = "RNA"
rep_number <- ""
if (grepl("rep_", reads_psite_sample)) {
  # Extract the number immediately following "rep"
  rep_number <- str_extract(reads_psite_sample, "(?<=rep_)\\d+")
  }
# Combine the genotype, rfp/rna, and rep number information into a new sample name.
sample_name = paste(genotype, rfp_or_rna, rep_number, sep = "_")

ggplot(counts_complete, aes(x = Position, y = ReadCount)) + geom_line() +
  labs(x = "position", y = "read count", title = paste(transcript_id, " read counts in ", sample_name), sep = "") +
  theme_bw() + 
  theme(panel.grid.minor = element_blank()) +
  coord_cartesian(ylim = c(0,0.15), xlim= c(0,2500))

#Limit max position to 5000 for ZNF562
#Limit max position to 500 for LLPH
#Limit max position to 350 for SNURF
#Limit max position to 300 for SEM1

# ggsave(filename = paste("output/20230722_hiND14/", transcript_id, "_", sample_name, ".png", sep =""), plot = p, width = 8, height = 4, dpi = 600)
```

Checking read length distribution.
It seemed that for wild type average, I have longer read length overall compared to Nterm average.
Nevertheless, the median for both is between 28-29.

```{r}
sample_list <- list("wt_avg" = c("hiND14_wt_rep_1_RFP", "hiND14_wt_rep_2_RFP", "hiND14_wt_rep_3_RFP"), 
                    "Nterm_avg" = c("hiND14_Nterm_rep_1_RFP", "hiND14_Nterm_rep_2_RFP", "hiND14_Nterm_rep_3_RFP"))

length_dist_dodged <- rlength_distr(bam_list, 
                                    sample = sample_list, cl = 99,
                                    multisamples = "average",
                                    plot_style = "dodged",
                                    colour = c("dodgerblue", "gray70"))
length_dist_dodged[["plot"]] 

```
Checking for trinucleotide periodicity.
Generally the P-site signal is a bit low..

```{r}
example_frames_stratified <- frame_psite_length(reads_psite_list, sample = "hiND14_Nterm_rep_3_RFP",
                                                region = "all", cl = 90)
example_frames_stratified[["plot"]]
```
Check for metaprofile psite as a representative of trinucleotide periodicity across all start codons.

```{r}
sample_list <- list("wt_avg" = c("hiND14_wt_rep_1_RFP", "hiND14_wt_rep_2_RFP", "hiND14_wt_rep_3_RFP"), 
                    "Nterm_avg" = c("hiND14_Nterm_rep_1_RFP", "hiND14_Nterm_rep_2_RFP", "hiND14_Nterm_rep_3_RFP"))
example_metaprofile_overlaid <- metaprofile_psite(reads_psite_list, annotation_dt, 
                                                  sample = sample_list, 
                                                  multisamples = "average", plot_style = "overlaid",
                                                  utr5l = 20, cdsl = 40, utr3l = 20,
                                                  frequency = TRUE, plot_title = "transcript",
                                                  colour = c("aquamarine4", "gray70"))
example_metaprofile_overlaid[["plot"]]
```

