---
title: "Vicky_codon_usage[20230601]"
output: html_notebook
---

Load all the required packages:
```{r}
library(tidyr)
library(dplyr)
library(stringr)
library(ggplot2)
library(riboWaltz)
library(data.table)
```

This notebook processes .bam files generated using STAR alignment against primary mouse GRCm39 GENCODE annotation. .bam files used are in ./data/Vicky

I am using riboWaltz to process this data.
I create annotation_dt using the human primary assembly basic annotation that I used to align my RFP data with STAR. I'm also loading the data and convert them into list.

```{r}
annotation_dt=create_annotation(gtfpath = "data/Vicky/gencode.vM32.primary_assembly.basic.annotation.gtf")
bam_list <- bamtolist("data/Vicky", annotation = annotation_dt, refseq_sep = "|")
```
For now, I'm not filtering the reads.

```{r}
# Try filter only read lengths that satisfy a periodicity threshold
# (i.e. 70% of the reads fall in one of the three reading frames along the CDS)
#filtered_bam_list <- length_filter(data = bam_list,
#                                   length_filter_mode = "periodicity",
#                                   periodicity_threshold = 70)
```

Call for psite, flanking = 6, extremity = "auto"
Use default recommended settings:
Also save the offsets per sample in a text file

```{r}
offsets <- psite(data = bam_list)
samples <- unique(offsets$sample)
for (i in 1:length(samples)) {
  sam_offs <- subset(offsets, sample == samples[i], 
                    select = c("length", "corrected_offset_from_3"))
  colnames(sam_offs) <- c("length", "p_offset")
  write.table(sam_offs, paste("output/Vicky/", samples[i], "_p-site-offsets.tsv", sep=""), 
              sep = "\t", row.names = FALSE)
}
```
Get the main table of reads with psite information.

```{r}
reads_psite_list <- psite_info(bam_list, offsets)
```

I'm curious about the periodicity. Looking at all samples in rep1:

```{r}
# rep1_A2H12_RFP_Aligned.toTranscriptome.out
# rep2_A1A11_RFP_Aligned.toTranscriptome.out
# rep3_A1C11_RFP_Aligned.toTranscriptome.out
# rep1_C1B1_RFP_Aligned.toTranscriptome.out
# rep2_C1B2_RFP_Aligned.toTranscriptome.out
# rep3_C1B4_RFP_Aligned.toTranscriptome.out
# rep1_E2E1_RFP_Aligned.toTranscriptome.out
# rep2_E2E10_RFP_Aligned.toTranscriptome.out
# rep3_E2B7_RFP_Aligned.toTranscriptome.out

rep1_A2H12_frames_stratified <- frame_psite_length(reads_psite_list, 
                                                   sample = "rep1_A2H12_RFP_Aligned.toTranscriptome.out",
                                                   region = "all", cl = 90)
rep1_A2H12_frames_stratified[["plot"]]
rep1_C1B1_frames_stratified <- frame_psite_length(reads_psite_list, 
                                                  sample = "rep1_C1B1_RFP_Aligned.toTranscriptome.out",
                                                  region = "all", cl = 90)
rep1_C1B1_frames_stratified[["plot"]]
rep1_E2E1_frames_stratified <- frame_psite_length(reads_psite_list,
                                                  sample = "rep1_E2E1_RFP_Aligned.toTranscriptome.out",
                                                  region = "all", cl = 90)
rep1_E2E1_frames_stratified[["plot"]]
```

Check for metaprofile psite as a representative of trinucleotide periodicity across all start codons.

```{r}
# rep1_A2H12_RFP_Aligned.toTranscriptome.out
# rep2_A1A11_RFP_Aligned.toTranscriptome.out
# rep3_A1C11_RFP_Aligned.toTranscriptome.out
# rep1_C1B1_RFP_Aligned.toTranscriptome.out
# rep2_C1B2_RFP_Aligned.toTranscriptome.out
# rep3_C1B4_RFP_Aligned.toTranscriptome.out
# rep1_E2E1_RFP_Aligned.toTranscriptome.out
# rep2_E2E10_RFP_Aligned.toTranscriptome.out
# rep3_E2B7_RFP_Aligned.toTranscriptome.out

Ctr_Ala_list <- list("Ctrl_avg" = c("rep1_C1B1_RFP_Aligned.toTranscriptome.out",
                                    "rep2_C1B2_RFP_Aligned.toTranscriptome.out", 
                                    "rep3_C1B4_RFP_Aligned.toTranscriptome.out"), 
                    "Ala_avg" = c("rep1_A2H12_RFP_Aligned.toTranscriptome.out", 
                                  "rep2_A1A11_RFP_Aligned.toTranscriptome.out", 
                                  "rep3_A1C11_RFP_Aligned.toTranscriptome.out"))
Ctr_Gln_list <- list("Ctrl_avg" = c("rep1_C1B1_RFP_Aligned.toTranscriptome.out",
                                    "rep2_C1B2_RFP_Aligned.toTranscriptome.out", 
                                    "rep3_C1B4_RFP_Aligned.toTranscriptome.out"), 
                    "Gln_avg" = c("rep1_E2E1_RFP_Aligned.toTranscriptome.out", 
                                  "rep2_E2E10_RFP_Aligned.toTranscriptome.out", 
                                  "rep3_E2B7_RFP_Aligned.toTranscriptome.out"))


Ctrl_Ala_metaprofile_overlaid <- metaprofile_psite(reads_psite_list, annotation_dt,
                                                   sample = Ctr_Ala_list,
                                                   multisamples = "average", plot_style = "overlaid",
                                                   utr5l = 20, cdsl = 40, utr3l = 20,
                                                   frequency = TRUE, plot_title = "transcript",
                                                   colour = c("aquamarine4", "gray70"))
Ctrl_Ala_metaprofile_overlaid[["plot"]]
Ctrl_Gln_metaprofile_overlaid <- metaprofile_psite(reads_psite_list, annotation_dt,
                                                   sample = Ctr_Gln_list,
                                                   multisamples = "average", plot_style = "overlaid",
                                                   utr5l = 20, cdsl = 40, utr3l = 20,
                                                   frequency = TRUE, plot_title = "transcript",
                                                   colour = c("aquamarine4", "gray70"))
Ctrl_Gln_metaprofile_overlaid[["plot"]]
```

To remove errors relating to p-site/a-site mapped to codons outside the annotation:
I'm filtering out all reads mapped to 3'utr, and those mapped right before the stop codon.

```{r}
for(n in names(reads_psite_list)) {
  reads_psite_list[[n]] <- reads_psite_list[[n]][!which(reads_psite_list[[n]]$psite_region == "3utr")]
  reads_psite_list[[n]] <- reads_psite_list[[n]][!which(reads_psite_list[[n]]$psite_region == "cds" & reads_psite_list[[n]]$end3 < reads_psite_list[[n]]$psite + 3)]
  reads_psite_list[[n]] <- reads_psite_list[[n]][!is.na(reads_psite_list[[n]]$psite_region)]
}
```

Checking for codon usage:
p-site

```{r}
# rep1_A2H12_RFP_Aligned.toTranscriptome.out
# rep2_A1A11_RFP_Aligned.toTranscriptome.out
# rep3_A1C11_RFP_Aligned.toTranscriptome.out
# rep1_C1B1_RFP_Aligned.toTranscriptome.out
# rep2_C1B2_RFP_Aligned.toTranscriptome.out
# rep3_C1B4_RFP_Aligned.toTranscriptome.out
# rep1_E2E1_RFP_Aligned.toTranscriptome.out
# rep2_E2E10_RFP_Aligned.toTranscriptome.out
# rep3_E2B7_RFP_Aligned.toTranscriptome.out

fasta_path = "data/Vicky/GRCm39.primary_assembly.genome.fa"
gtf_path = "data/Vicky/gencode.vM32.primary_assembly.basic.annotation.gtf"
codon_table <- read.table(file = "codon_table.tsv", sep = "\t", header=TRUE)

codon_usage_psite_all <- codon_table
codon_rank_psite_all <- codon_table

for(n in names(reads_psite_list)) {
  codon_usage <- codon_usage_psite(reads_psite_list, annotation = annotation_dt, 
                                   sample = n, fastapath = fasta_path, gtfpath = gtf_path, 
                                   fasta_genome = TRUE, refseq_sep = " ", site = "psite")
  codon_usage_psite_all <- left_join(codon_usage_psite_all,
                                     codon_usage$dt[, .(codon, normalized_value)], by="codon")
  normalized_value_index <- which(colnames(codon_usage_psite_all) == "normalized_value")
  colnames(codon_usage_psite_all)[normalized_value_index] <- n
  codon_rank_psite_all[[n]] <- rank(codon_usage_psite_all[[n]])
}

codon_usage_psite_all$wt_RFP_ave = rowMeans(
  codon_usage_psite_all[, c("rep1_C1B1_RFP_Aligned.toTranscriptome.out",
                            "rep2_C1B2_RFP_Aligned.toTranscriptome.out",
                            "rep3_C1B4_RFP_Aligned.toTranscriptome.out")])

codon_usage_psite_all$Ala_RFP_ave = rowMeans(
  codon_usage_psite_all[, c("rep1_A2H12_RFP_Aligned.toTranscriptome.out", 
                            "rep2_A1A11_RFP_Aligned.toTranscriptome.out", 
                            "rep3_A1C11_RFP_Aligned.toTranscriptome.out")])

codon_usage_psite_all$Gln_RFP_ave = rowMeans(
  codon_usage_psite_all[, c("rep1_E2E1_RFP_Aligned.toTranscriptome.out", 
                            "rep2_E2E10_RFP_Aligned.toTranscriptome.out", 
                            "rep3_E2B7_RFP_Aligned.toTranscriptome.out")])

codon_usage_psite_all <- codon_usage_psite_all %>%
  mutate(wt_RFP_ave_plot = wt_RFP_ave/max(codon_usage_psite_all$wt_RFP_ave),
         Ala_RFP_ave_plot = Ala_RFP_ave/max(codon_usage_psite_all$Ala_RFP_ave),
         Gln_RFP_ave_plot = Gln_RFP_ave/max(codon_usage_psite_all$Gln_RFP_ave))
```

Plot Ala vs ctrl p-site codon usage

```{r}
bs <- 25
label_number <- 0
plot_df <- codon_usage_psite_all
#plot_df <- codon_usage_asite_all
setDT(plot_df)
plot_df <- plot_df[order(wt_RFP_ave_plot)][, `:=`(codon, factor(codon, levels = codon))]

correlation <- round(cor(plot_df$wt_RFP_ave_plot, 
                         plot_df$Ala_RFP_ave_plot), 3)

fit <- lm(plot_df$Ala_RFP_ave_plot ~ plot_df$wt_RFP_ave_plot)
outlier_tab <- data.table(predict.lm(fit, interval = "confidence", level = 0.95))[, `:=`(codon, plot_df$codon)][, `:=`(Ala_RFP_ave_plot, plot_df$Ala_RFP_ave_plot)][Ala_RFP_ave_plot < lwr | Ala_RFP_ave_plot > upr]
outlier_cod <- as.character(outlier_tab[, `:=`(max_dist, min(abs(Ala_RFP_ave_plot - lwr), 
                                                             abs(Ala_RFP_ave_plot - upr))), 
                                        by = 1:nrow(outlier_tab)][order(-max_dist)][1:label_number, codon])
lab_table <- plot_df[, `:=`(codon_out, "")][as.character(codon) %in% outlier_cod, `:=`(codon_out, codon)
                                               ][, `:=`(aa_out, "")][codon_out != "", `:=`(aa_out, aa)]

ggplot(plot_df, aes(x = wt_RFP_ave_plot, y=Ala_RFP_ave_plot)) +
  geom_smooth(method = "lm", se = TRUE, color = "gray80", fill = "gray80", linetype= 1,
              formula = y ~ x, level = 0.95, fullrange = TRUE) +
  geom_point(alpha = 0.5, size = bs* 0.14) +
  theme_bw(base_size = bs) +
  theme(legend.position = "top", legend.margin = margin(0,0,0,0),
        legend.box.margin = margin(5,0,-15,0)) +
  theme(legend.text = element_text(margin = margin(l = -12, unit = "pt"))) +
  theme(panel.grid.minor.y = element_blank(), panel.grid.minor.x = element_blank()) +
  scale_x_continuous(name = "wt", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  scale_y_continuous(name = "Ala", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  coord_cartesian(xlim = c(-0.05, 1.05), ylim = c(-0.05, 1.05)) +
  annotate("text", x = 1, y = 0, label = paste0("R=", correlation), 
           vjust = -0.2, size = bs * 0.2, hjust = 1, color = "black") +
  geom_abline(intercept = 0, slope = 1, color = "red") +
  ggrepel::geom_text_repel(data = lab_table, aes(label = codon_out), show.legend = F)
```

Plot Gln vs ctrl p-site codon usage

```{r}
bs <- 25
label_number <- 0
plot_df <- codon_usage_psite_all
#plot_df <- codon_usage_asite_all
setDT(plot_df)
plot_df <- plot_df[order(wt_RFP_ave_plot)][, `:=`(codon, factor(codon, levels = codon))]

correlation <- round(cor(plot_df$wt_RFP_ave_plot, 
                         plot_df$Gln_RFP_ave_plot), 3)

fit <- lm(plot_df$Gln_RFP_ave_plot ~ plot_df$wt_RFP_ave_plot)
outlier_tab <- data.table(predict.lm(fit, interval = "confidence", level = 0.95))[, `:=`(codon, plot_df$codon)][, `:=`(Gln_RFP_ave_plot, plot_df$Gln_RFP_ave_plot)][Gln_RFP_ave_plot < lwr | Gln_RFP_ave_plot > upr]
outlier_cod <- as.character(outlier_tab[, `:=`(max_dist, min(abs(Gln_RFP_ave_plot - lwr), 
                                                             abs(Gln_RFP_ave_plot - upr))), 
                                        by = 1:nrow(outlier_tab)][order(-max_dist)][1:label_number, codon])
lab_table <- plot_df[, `:=`(codon_out, "")][as.character(codon) %in% outlier_cod, `:=`(codon_out, codon)
                                               ][, `:=`(aa_out, "")][codon_out != "", `:=`(aa_out, aa)]

ggplot(plot_df, aes(x = wt_RFP_ave_plot, y=Gln_RFP_ave_plot)) +
  geom_smooth(method = "lm", se = TRUE, color = "gray80", fill = "gray80", linetype= 1,
              formula = y ~ x, level = 0.95, fullrange = TRUE) +
  geom_point(alpha = 0.5, size = bs* 0.14) +
  theme_bw(base_size = bs) +
  theme(legend.position = "top", legend.margin = margin(0,0,0,0),
        legend.box.margin = margin(5,0,-15,0)) +
  theme(legend.text = element_text(margin = margin(l = -12, unit = "pt"))) +
  theme(panel.grid.minor.y = element_blank(), panel.grid.minor.x = element_blank()) +
  scale_x_continuous(name = "wt", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  scale_y_continuous(name = "Gln", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  coord_cartesian(xlim = c(-0.05, 1.05), ylim = c(-0.05, 1.05)) +
  annotate("text", x = 1, y = 0, label = paste0("R=", correlation), 
           vjust = -0.2, size = bs * 0.2, hjust = 1, color = "black") +
  geom_abline(intercept = 0, slope = 1, color = "red") +
  ggrepel::geom_text_repel(data = lab_table, aes(label = codon_out), show.legend = F)
```

Checking for codon usage:
a-site

```{r}
# rep1_A2H12_RFP_Aligned.toTranscriptome.out
# rep2_A1A11_RFP_Aligned.toTranscriptome.out
# rep3_A1C11_RFP_Aligned.toTranscriptome.out
# rep1_C1B1_RFP_Aligned.toTranscriptome.out
# rep2_C1B2_RFP_Aligned.toTranscriptome.out
# rep3_C1B4_RFP_Aligned.toTranscriptome.out
# rep1_E2E1_RFP_Aligned.toTranscriptome.out
# rep2_E2E10_RFP_Aligned.toTranscriptome.out
# rep3_E2B7_RFP_Aligned.toTranscriptome.out

fasta_path = "data/Vicky/GRCm39.primary_assembly.genome.fa"
gtf_path = "data/Vicky/gencode.vM32.primary_assembly.basic.annotation.gtf"
codon_table <- read.table(file = "codon_table.tsv", sep = "\t", header=TRUE)

codon_usage_asite_all <- codon_table
codon_rank_asite_all <- codon_table

for(n in names(reads_psite_list)) {
  codon_usage <- codon_usage_psite(reads_psite_list, annotation = annotation_dt, 
                                   sample = n, fastapath = fasta_path, gtfpath = gtf_path, 
                                   fasta_genome = TRUE, refseq_sep = " ", site = "asite")
  codon_usage_asite_all <- left_join(codon_usage_asite_all,
                                     codon_usage$dt[, .(codon, normalized_value)], by="codon")
  normalized_value_index <- which(colnames(codon_usage_asite_all) == "normalized_value")
  colnames(codon_usage_asite_all)[normalized_value_index] <- n
  codon_rank_psite_all[[n]] <- rank(codon_usage_asite_all[[n]])
}

codon_usage_asite_all$wt_RFP_ave = rowMeans(
  codon_usage_asite_all[, c("rep1_C1B1_RFP_Aligned.toTranscriptome.out",
                            "rep2_C1B2_RFP_Aligned.toTranscriptome.out",
                            "rep3_C1B4_RFP_Aligned.toTranscriptome.out")])

codon_usage_asite_all$Ala_RFP_ave = rowMeans(
  codon_usage_asite_all[, c("rep1_A2H12_RFP_Aligned.toTranscriptome.out", 
                            "rep2_A1A11_RFP_Aligned.toTranscriptome.out", 
                            "rep3_A1C11_RFP_Aligned.toTranscriptome.out")])

codon_usage_asite_all$Gln_RFP_ave = rowMeans(
  codon_usage_asite_all[, c("rep1_E2E1_RFP_Aligned.toTranscriptome.out", 
                            "rep2_E2E10_RFP_Aligned.toTranscriptome.out", 
                            "rep3_E2B7_RFP_Aligned.toTranscriptome.out")])

codon_usage_asite_all <- codon_usage_asite_all %>%
  mutate(wt_RFP_ave_plot = wt_RFP_ave/max(codon_usage_asite_all$wt_RFP_ave),
         Ala_RFP_ave_plot = Ala_RFP_ave/max(codon_usage_asite_all$Ala_RFP_ave),
         Gln_RFP_ave_plot = Gln_RFP_ave/max(codon_usage_asite_all$Gln_RFP_ave))
```

Plot Ala vs ctrl a-site codon usage

```{r}
bs <- 25
label_number <- 0
#plot_df <- codon_usage_psite_all
plot_df <- codon_usage_asite_all
setDT(plot_df)
plot_df <- plot_df[order(wt_RFP_ave_plot)][, `:=`(codon, factor(codon, levels = codon))]

correlation <- round(cor(plot_df$wt_RFP_ave_plot, 
                         plot_df$Ala_RFP_ave_plot), 3)

fit <- lm(plot_df$Ala_RFP_ave_plot ~ plot_df$wt_RFP_ave_plot)
outlier_tab <- data.table(predict.lm(fit, interval = "confidence", level = 0.95))[, `:=`(codon, plot_df$codon)][, `:=`(Ala_RFP_ave_plot, plot_df$Ala_RFP_ave_plot)][Ala_RFP_ave_plot < lwr | Ala_RFP_ave_plot > upr]
outlier_cod <- as.character(outlier_tab[, `:=`(max_dist, min(abs(Ala_RFP_ave_plot - lwr), 
                                                             abs(Ala_RFP_ave_plot - upr))), 
                                        by = 1:nrow(outlier_tab)][order(-max_dist)][1:label_number, codon])
lab_table <- plot_df[, `:=`(codon_out, "")][as.character(codon) %in% outlier_cod, `:=`(codon_out, codon)
                                               ][, `:=`(aa_out, "")][codon_out != "", `:=`(aa_out, aa)]

ggplot(plot_df, aes(x = wt_RFP_ave_plot, y=Ala_RFP_ave_plot)) +
  geom_smooth(method = "lm", se = TRUE, color = "gray80", fill = "gray80", linetype= 1,
              formula = y ~ x, level = 0.95, fullrange = TRUE) +
  geom_point(alpha = 0.5, size = bs* 0.14) +
  theme_bw(base_size = bs) +
  theme(legend.position = "top", legend.margin = margin(0,0,0,0),
        legend.box.margin = margin(5,0,-15,0)) +
  theme(legend.text = element_text(margin = margin(l = -12, unit = "pt"))) +
  theme(panel.grid.minor.y = element_blank(), panel.grid.minor.x = element_blank()) +
  scale_x_continuous(name = "wt", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  scale_y_continuous(name = "Ala", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  coord_cartesian(xlim = c(-0.05, 1.05), ylim = c(-0.05, 1.05)) +
  annotate("text", x = 1, y = 0, label = paste0("R=", correlation), 
           vjust = -0.2, size = bs * 0.2, hjust = 1, color = "black") +
  geom_abline(intercept = 0, slope = 1, color = "red") +
  ggrepel::geom_text_repel(data = lab_table, aes(label = codon_out), show.legend = F)
```

Plot Gln vs ctrl a-site codon usage

```{r}
bs <- 25
label_number <- 0
#plot_df <- codon_usage_psite_all
plot_df <- codon_usage_asite_all
setDT(plot_df)
plot_df <- plot_df[order(wt_RFP_ave_plot)][, `:=`(codon, factor(codon, levels = codon))]

correlation <- round(cor(plot_df$wt_RFP_ave_plot, 
                         plot_df$Gln_RFP_ave_plot), 3)

fit <- lm(plot_df$Gln_RFP_ave_plot ~ plot_df$wt_RFP_ave_plot)
outlier_tab <- data.table(predict.lm(fit, interval = "confidence", level = 0.95))[, `:=`(codon, plot_df$codon)][, `:=`(Gln_RFP_ave_plot, plot_df$Gln_RFP_ave_plot)][Gln_RFP_ave_plot < lwr | Gln_RFP_ave_plot > upr]
outlier_cod <- as.character(outlier_tab[, `:=`(max_dist, min(abs(Gln_RFP_ave_plot - lwr), 
                                                             abs(Gln_RFP_ave_plot - upr))), 
                                        by = 1:nrow(outlier_tab)][order(-max_dist)][1:label_number, codon])
lab_table <- plot_df[, `:=`(codon_out, "")][as.character(codon) %in% outlier_cod, `:=`(codon_out, codon)
                                               ][, `:=`(aa_out, "")][codon_out != "", `:=`(aa_out, aa)]

ggplot(plot_df, aes(x = wt_RFP_ave_plot, y=Gln_RFP_ave_plot)) +
  geom_smooth(method = "lm", se = TRUE, color = "gray80", fill = "gray80", linetype= 1,
              formula = y ~ x, level = 0.95, fullrange = TRUE) +
  geom_point(alpha = 0.5, size = bs* 0.14) +
  theme_bw(base_size = bs) +
  theme(legend.position = "top", legend.margin = margin(0,0,0,0),
        legend.box.margin = margin(5,0,-15,0)) +
  theme(legend.text = element_text(margin = margin(l = -12, unit = "pt"))) +
  theme(panel.grid.minor.y = element_blank(), panel.grid.minor.x = element_blank()) +
  scale_x_continuous(name = "wt", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  scale_y_continuous(name = "Gln", limits = c(-0.3, 1.3), 
                     breaks = c(0, 0.25, 0.5, 0.75, 1), expand = c(0, 0)) +
  coord_cartesian(xlim = c(-0.05, 1.05), ylim = c(-0.05, 1.05)) +
  annotate("text", x = 1, y = 0, label = paste0("R=", correlation), 
           vjust = -0.2, size = bs * 0.2, hjust = 1, color = "black") +
  geom_abline(intercept = 0, slope = 1, color = "red") +
  ggrepel::geom_text_repel(data = lab_table, aes(label = codon_out), show.legend = F)
```
